akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -1196,13 +1300,40 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
  * An asynchronous callback holder that is attached to a [[GraphStageLogic]].
  * Invoking [[AsyncCallback#invoke]] will eventually lead to the registered handler
  * being called.
+ *
+ * This holder has the same lifecycle as a stream and cannot be used before
+ * materialization is done.
+ *
+ * Typical use case is exchanging messages between stream and substreams
  */
 trait AsyncCallback[T] {
   /**
+   * Dispatch an asynchronous notification. This method must be thread-safe.
+   */
+  def invoke(t: T): Unit
+}
+
+/**
+ * This allows to add promise to async event (see [[AsyncCallbackAsMaterializedValue]])
+ * Promise will be failed once it's clear that event will not be proceed due to stream completion.
+ */
+trait HasCallbackPromise {
+  def done: Promise[_]
+}
+
+/**
+ * Version of asynchronous callback holder that may be invoked from external execution contexts.
+ * This callback holder can be used from as or within materialized value
+ */
+trait AsyncCallbackAsMaterializedValue[T] extends AsyncCallback[T] {
+
+  /**
    * Dispatch an asynchronous notification. This method is thread-safe and
    * may be invoked from external execution contexts.
+   * Promise in `HasCallbackPromise` will fail if stream is already closed or closed before
+   * being able to process the event
    */
-  def invoke(t: T): Unit
+  def invokeWithPromise[C <: T with HasCallbackPromise](event: C): Unit
```

Comment:
I agree, that would be simpler.

Reply:
@patriknw , @jrudolph In this case it will be  Future chain. So Futurecompletion that invokeWithPromise returns will complete another feature with specific response (like a number of bytes sent).
If project has success cases of using future chaining  or you feel this will be appropriate - let me fix it in this way